### PR TITLE
Feat/12 cstore scp

### PR DESCRIPTION
**Pull request by wayfarer3130**
Add an initial implementation of the c-store request.  Depends on an updated version of the dicom-parser library.

### CODE REVIEW

File: bin/dicomwebscp.js [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    const contexts = association.getPresentationContexts();
+    contexts.forEach((c) => {
+      const context = association.getPresentationContext(c.id);
+      if (
+        context.getAbstractSyntaxUid() === SopClass.Verification ||
+        Object.values(StorageClass).includes(context.getAbstractSyntaxUid())
+      ) {
+        const transferSyntaxes = context.getTransferSyntaxUids();
+        transferSyntaxes.forEach((transferSyntax) => {
+          if (AcceptedTransferSyntax[transferSyntax]) {
```
Comment (ladeirarodolfo) [inline bin/dicomwebscp.js]: What about not accepted transferSyntax? Maybe we should log something...

File: src/index.js [hunk 2]
```diff
@@ -12,171 +12,179 @@ const dirScanner = require('./dirScanner')
...
 const homedir = require('os').homedir();
+
 const dicomwebDefaultDir = '~/dicomweb';
 const Stats = require('./stats');
 const VideoWriter = require('./VideoWriter')
+const OverallStats = new Stats('OverallStats', 'Overall statistics');
+const StudyStats = new Stats('StudyStats', 'Study Generation', OverallStats);
+const handleHomeRelative = dirName => dirName[0] == '~' ? (path.join(homedir, dirName.substring(1))) : dirName;
+
+class StaticWado {
```
Comment (ladeirarodolfo) [inline src/index.js:26]: we might have conflicts with your PR and mine. the first in, second should fix conflicts

File: bin/dicomwebscp.js [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+  // Handle incoming C-STORE requests
+  cStoreRequest(request, callback) {
+    const importDs = request.getDataset().getDenaturalizedDataset();
+    const response = CStoreResponse.fromRequest(request);
+    const params = {TransferSyntaxUID: request.commandDataset.transferSyntaxUid};
+    
+    this.importer.importBinaryDicom(importDs, params).then(value => {
+      response.setStatus(Status.Success);
+      Stats.StudyStats.add('Receive DICOM', `Receive DICOM instance`, 250);
```
Comment (ladeirarodolfo) [inline bin/dicomwebscp.js]: default value for messageCount is actually 250 so the value is not need

File: src/index.js [hunk 4]
```diff
@@ -12,171 +12,179 @@ const dirScanner = require('./dirScanner')
...
+    static main(defaults) {
+        const importer = new StaticWado(defaults);
+        return importer.main();
     }
-    await callback.completeStudy();
-    OverallStats.summarize('Completed Study Processing');
 }
 
-dicomp10todicomweb.main = main;
+new StaticWado({});
```
Comment (ladeirarodolfo) [inline src/index.js]: is this line needed?

**Review by ladeirarodolfo — DISMISSED** [review]

**Comment by ladeirarodolfo** [comment]
@wayfarer3130  add few comments or request.
As far as the description (Add an initial implementation) is concerned, what I have seen looks good to me. We can improve, specially the server side of things once more request changes comes.

File: src/index.js (hunk 2 above)
Comment (wayfarer3130) [reply src/index.js:26]: Always an issue with multiple developers.

File: bin/dicomwebscp.js (hunk 1 above)
Comment (wayfarer3130) [reply bin/dicomwebscp.js]: No, logging something would leave a gazillion logs.  This is the wrong way to do this - the right way is to look to see if the most desired transfer syntax is available, and then proceed down the list, then if nothing is available, THEN log a message.  Going to tweak this to work properly

Comment (wayfarer3130) [reply bin/dicomwebscp.js]: Fixed this by choosing the desired tsuids in order, and then writing a message only on failure.

File: src/index.js (hunk 4 above)
Comment (wayfarer3130) [reply src/index.js]: No, was having failures on create in an odd spot, and wanted ot test the constructor.  Forgot to remote it.

File: bin/dicomwebscp.js (hunk 3 above)
Comment (wayfarer3130) [reply bin/dicomwebscp.js]: Fixed

**Pull request merged**